JAVA InstantiationException with mediainfo

Hi,

FITS seems to run as expected on a JPEG expect several JAVA errors appear that all seem to relate to mediainfo. It appears that Mediainfo is installed correctly.
I am using Windows 10, and the JAVA version is:
```
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
```
Is this an error on our end or is there some issue with FITS? Can we fix it?

```
Microsoft Windows [Version 10.0.17134.765]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\Users\Maggie>cd C:\Users\Maggie\Desktop\DCTools\fits-1.4.1


C:\Users\Maggie\Desktop\DCTools\fits-1.4.1>fits.bat -i C:\Users\Maggie\Desktop\DCTools\fits-1.4.1\Ex4-fits\20181231_183434.jpg
2019-05-29 09:14:41 - ERROR - ToolBelt:140 - Thread 1 error initializing edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo: java.lang.InstantiationException  Message: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo
java.lang.InstantiationException: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo
        at java.lang.Class.newInstance(Unknown Source)
        at edu.harvard.hul.ois.fits.tools.ToolBelt.createToolClassInstance(ToolBelt.java:174)
        at edu.harvard.hul.ois.fits.tools.ToolBelt.init(ToolBelt.java:122)
        at edu.harvard.hul.ois.fits.tools.ToolBelt.<init>(ToolBelt.java:58)
        at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:243)
        at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:123)
        at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:112)
        at edu.harvard.hul.ois.fits.Fits.constructFits(Fits.java:325)
        at edu.harvard.hul.ois.fits.Fits.main(Fits.java:307)
Caused by: java.lang.NoSuchMethodException: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo.<init>()
        at java.lang.Class.getConstructor0(Unknown Source)
        ... 9 more
<?xml version="1.0" encoding="UTF-8"?>
<fits xmlns="http://hul.harvard.edu/ois/xml/ns/fits/fits_output" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://hul.harvard.edu/ois/xml/ns/fits/fits_output http://hul.harvard.edu/ois/xml/xsd/fits/fits_output.xsd" version="1.4.1" timestamp="5/29/19 9:14 AM">
  <identification>
    <identity format="JPEG EXIF" mimetype="image/jpeg" toolname="FITS" toolversion="1.4.1">
      <tool toolname="Exiftool" toolversion="11.14" />
      <tool toolname="NLNZ Metadata Extractor" toolversion="3.6GA" />
      <version toolname="Exiftool" toolversion="11.14">1.01</version>
    </identity>
  </identification>
  <fileinfo>
    <size toolname="Jhove" toolversion="1.20.1">329338</size>
    <creatingApplicationName toolname="Exiftool" toolversion="11.14" status="CONFLICT">SM-G950U</creatingApplicationName>
    <creatingApplicationName toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">G950USQU5CRK1</creatingApplicationName>
    <lastmodified toolname="Exiftool" toolversion="11.14" status="CONFLICT">2018:12:31 18:34:33</lastmodified>
    <lastmodified toolname="Tika" toolversion="1.19.1" status="CONFLICT">2018-12-31T18:34:33</lastmodified>
    <created toolname="Exiftool" toolversion="11.14">2018:12:31 18:34:33</created>
    <filepath toolname="OIS File Information" toolversion="0.2" status="SINGLE_RESULT">C:\Users\Maggie\Desktop\DCTools\fits-1.4.1\Ex4-fits\20181231_183434.jpg</filepath>
    <filename toolname="OIS File Information" toolversion="0.2" status="SINGLE_RESULT">20181231_183434.jpg</filename>
    <md5checksum toolname="OIS File Information" toolversion="0.2" status="SINGLE_RESULT">2333a84077be1149c3b4dbc89442743e</md5checksum>
    <fslastmodified toolname="OIS File Information" toolversion="0.2" status="SINGLE_RESULT">1555666061851</fslastmodified>
  </fileinfo>
  <filestatus />
  <metadata>
    <image>
      <imageWidth toolname="Exiftool" toolversion="11.14">3264</imageWidth>
      <imageHeight toolname="Exiftool" toolversion="11.14">2448</imageHeight>
      <YCbCrSubSampling toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">2 2</YCbCrSubSampling>
      <YCbCrPositioning toolname="Exiftool" toolversion="11.14">1</YCbCrPositioning>
      <orientation toolname="Exiftool" toolversion="11.14">normal, rotated 180┬░</orientation>
      <samplingFrequencyUnit toolname="Exiftool" toolversion="11.14" status="CONFLICT">no absolute unit of measurement</samplingFrequencyUnit>
      <samplingFrequencyUnit toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">in.</samplingFrequencyUnit>
      <xSamplingFrequency toolname="Exiftool" toolversion="11.14" status="CONFLICT">1</xSamplingFrequency>
      <xSamplingFrequency toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">72</xSamplingFrequency>
      <ySamplingFrequency toolname="Exiftool" toolversion="11.14" status="CONFLICT">1</ySamplingFrequency>
      <ySamplingFrequency toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">72</ySamplingFrequency>
      <bitsPerSample toolname="Exiftool" toolversion="11.14">8 8 8</bitsPerSample>
      <scannerManufacturer toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">samsung</scannerManufacturer>
      <scannerModelName toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">SM-G950U</scannerModelName>
      <scanningSoftwareName toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">G950USQU5CRK1</scanningSoftwareName>
      <fNumber toolname="Exiftool" toolversion="11.14">1.7</fNumber>
      <exposureTime toolname="Exiftool" toolversion="11.14" status="CONFLICT">0.0769</exposureTime>
      <exposureTime toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">0.07692307692307693</exposureTime>
      <exposureProgram toolname="Exiftool" toolversion="11.14">Normal program</exposureProgram>
      <isoSpeedRating toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">500</isoSpeedRating>
      <exifVersion toolname="Exiftool" toolversion="11.14">0220</exifVersion>
      <maxApertureValue toolname="Exiftool" toolversion="11.14" status="CONFLICT">1.7</maxApertureValue>
      <maxApertureValue toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">1.53</maxApertureValue>
      <meteringMode toolname="Exiftool" toolversion="11.14">Center weighted average</meteringMode>
      <lightSource toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="SINGLE_RESULT">unknown</lightSource>
      <flash toolname="Exiftool" toolversion="11.14">Flash did not fire</flash>
      <focalLength toolname="Exiftool" toolversion="11.14" status="CONFLICT">3.0</focalLength>
      <focalLength toolname="NLNZ Metadata Extractor" toolversion="3.6GA" status="CONFLICT">2.95</focalLength>
      <gpsLatitudeRef toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">N</gpsLatitudeRef>
      <gpsLatitude toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">38 deg 54' 0.00" N</gpsLatitude>
      <gpsLongitudeRef toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">W</gpsLongitudeRef>
      <gpsLongitude toolname="Exiftool" toolversion="11.14" status="SINGLE_RESULT">77 deg 1' 37.00" W</gpsLongitude>
    </image>
  </metadata>
  <statistics fitsExecutionTime="694">
    <tool toolname="MediaInfo" toolversion="[unknown]" status="failed" />
    <tool toolname="OIS Audio Information" toolversion="0.1" status="did not run" />
    <tool toolname="ADL Tool" toolversion="0.1" status="did not run" />
    <tool toolname="VTT Tool" toolversion="0.1" status="did not run" />
    <tool toolname="Droid" toolversion="6.4" executionTime="128" />
    <tool toolname="Jhove" toolversion="1.20.1" executionTime="658" />
    <tool toolname="file utility" toolversion="5.03" executionTime="627" />
    <tool toolname="Exiftool" toolversion="11.14" executionTime="638" />
    <tool toolname="NLNZ Metadata Extractor" toolversion="3.6GA" executionTime="612" />
    <tool toolname="OIS File Information" toolversion="0.2" executionTime="127" />
    <tool toolname="OIS XML Metadata" toolversion="0.2" status="did not run" />
    <tool toolname="ffident" toolversion="0.2" executionTime="563" />
    <tool toolname="Tika" toolversion="1.19.1" executionTime="223" />
  </statistics>
</fits>


C:\Users\Maggie\Desktop\DCTools\fits-1.4.1>

C:\Users\Maggie\Desktop\DCTools\fits-1.4.1>java -version
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
Java HotSpot(TM) Client VM (build 25.211-b12, mixed mode)

C:\Users\Maggie\Desktop\DCTools\fits-1.4.1>
```

I sometimes see the same error on linux:

```
2021-07-01 13:45:50 - ERROR - MediaInfo:95 - Error loading native library for this operating system for tool: MediaInfo. ostype=[Linux] -- jvmModel=[64] -- nativeLibPath=[current/tools/mediainfo/linux] -- No native MediaInfo library for this OS
java.lang.UnsatisfiedLinkError: Unable to load library 'mediainfo': Native library (linux-x86-64/libmediainfo.so) not found in resource path (:current/lib/aes-31.3.jar:current/lib/commons-cli-1.2.jar:current/lib/commons-collections-3.2.2.jar:current/lib/commons-configuration-1.8.jar:current/lib/commons-io-2.4.jar:current/lib/commons-lang-2.6.jar:current/lib/commons-logging-1.2.jar:current/lib/fast-md5-2.7.1.jar:current/lib/fits-1.5.0.jar:current/lib/gson-2.4.jar:current/lib/jaxen-1.1.6.jar:current/lib/jdom-1.1.3.jar:current/lib/jna-4.1.0.jar:current/lib/log4j-1.2.17.jar:current/lib/ots-1.0.53.jar:current/lib/Saxon-HE-9.8.0-10.jar:current/lib/slf4j-api-1.7.22.jar:current/lib/slf4j-log4j12-1.7.22.jar:current/lib/stax2-api-3.1.4.jar:current/lib/staxmate-2.0.1.jar:current/lib/woodstox-core-lgpl-4.4.1.jar:current/lib/xercesImpl-2.9.1.jar)
	at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:271)
	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:398)
	at com.sun.jna.Library$Handler.<init>(Library.java:147)
	at com.sun.jna.Native.loadLibrary(Native.java:412)
	at edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfoNativeWrapper$MediaInfoDLL_Internal.<clinit>(MediaInfoNativeWrapper.java:101)
	at edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfoNativeWrapper.Option_Static(MediaInfoNativeWrapper.java:442)
	at edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo.<init>(MediaInfo.java:86)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at edu.harvard.hul.ois.fits.tools.ToolBelt.createToolClassInstance(ToolBelt.java:169)
	at edu.harvard.hul.ois.fits.tools.ToolBelt.init(ToolBelt.java:122)
	at edu.harvard.hul.ois.fits.tools.ToolBelt.<init>(ToolBelt.java:58)
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:244)
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:124)
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:113)
	at edu.harvard.hul.ois.fits.Fits.constructFits(Fits.java:348)
	at edu.harvard.hul.ois.fits.Fits.main(Fits.java:326)
2021-07-01 13:45:50 - ERROR - ToolBelt:139 - Thread [1] Error instantiating class: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo -- Exception thrown: java.lang.InstantiationException -- Error message: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo
2021-07-01 13:45:54 - ERROR - Fits:677 - Tool error processing file: TECH0
edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo
java.lang.InstantiationException: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo
	at java.base/java.lang.Class.newInstance(Class.java:571)
	at edu.harvard.hul.ois.fits.tools.ToolBelt.createToolClassInstance(ToolBelt.java:174)
	at edu.harvard.hul.ois.fits.tools.ToolBelt.init(ToolBelt.java:122)
	at edu.harvard.hul.ois.fits.tools.ToolBelt.<init>(ToolBelt.java:58)
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:244)
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:124)
	at edu.harvard.hul.ois.fits.Fits.<init>(Fits.java:113)
	at edu.harvard.hul.ois.fits.Fits.constructFits(Fits.java:348)
	at edu.harvard.hul.ois.fits.Fits.main(Fits.java:326)
Caused by: java.lang.NoSuchMethodException: edu.harvard.hul.ois.fits.tools.mediainfo.MediaInfo.<init>()
	at java.base/java.lang.Class.getConstructor0(Class.java:3349)
	at java.base/java.lang.Class.newInstance(Class.java:556)
	... 8 more

```

I originally thought that it was a problem with how the library was being loaded, but it turned out that the error is extremely inconsistent, which seems to suggest that there's some non-deterministic behavior in the class loading that is causing a problem, noting that the native lib is loaded in a class level static block.

--- a/src/main/java/edu/harvard/hul/ois/fits/tools/mediainfo/MediaInfoNativeWrapper.java
+++ b/src/main/java/edu/harvard/hul/ois/fits/tools/mediainfo/MediaInfoNativeWrapper.java
@@ -60,7 +60,7 @@ class MediaInfoNativeWrapper
                final String LocalPath;
                if (loader!=null)
                {
-                   LocalPath=loader.getResource(MediaInfoNativeWrapper.class.getName().replace('.', '/')+ ".class").getPath().replace("MediaInfo.class", "");
+                   LocalPath=loader.getResource(MediaInfoNativeWrapper.class.getName().replace('.', '/')+ ".class").getPath().replace("MediaInfoNativeWrapper.class", "");
                    try
                    {
                        NativeLibrary.getInstance(LocalPath+"libzen.so.0"); // Local path
@@ -84,6 +84,7 @@ class MediaInfoNativeWrapper
                    }
                    catch (LinkageError e)
                    {
+                       NativeLibrary.getInstance("mediainfo");
                    }
                }
            }

